proposals/0014-wave-size-range.md

```diff
@@ -0,0 +1,283 @@
+
+To make use of this attribute, SPIRV would have to represent it in new SPIRV
+ operations that drivers can read to determine what the values are an how to
+ use to them to choose the best wave size for a given shader.
+
+### Diagnostic Changes
+
+#### New Errors
+
+These are where new or slightly altered errors are produced:
+
+* If any of the parameters of `WaveSize` are not literal power of two integers
+  between 4 and 128 inclusive.
+* If the minimum wave size value is greater than or equal to the max wave size
+ as respectively specified by the first and second parameters in the `WaveSize`
+ attribute.
+* If the preferred wave size value indicated by the third parameter of
+ `WaveSize` is not in the range specified by the first two parameters.
+* If multiple `WaveSize` attributes are applied to the same entry point,
+ regardless of what overload they are, the existing error indicating an
+ attribute conflict is produced.
+* If more than three or fewer than one parameter is applied to `WaveSize`.
+* If negative values are provided for any of the `WaveSize` parameters.
+* If float values are provided for any of the `WaveSize` parameters.
+* If non-numerical values are provided for any of the `WaveSize` parameters.
+* If non-literal variables are provided for any of the `WaveSize` parameters.
+* If any of the parameters to `WaveSize` are less than 4 or greater than 128.
+
+#### Validation Changes
+
+Validation should confirm:
+
+* The tuple that the wave size range tag points to has exactly three elements.
+* Each element in that is a power of two integer between 4 and 128 inclusive.
+  The third parameter may also be zero.
+* The minimum wave size is less than the maximum wave size.
+* The preferred wave size lies between the minimum and maximum.
+
+### Runtime Additions
+
+#### Runtime information
+
+The PSV0 runtime data structure already contains both
+ `MinimumExpectedWaveLaneCount` and `MaximumExpectedWaveLaneCount` members that
+ can be used to transmit the minimum and maximum values to the runtime.
+A new `PreferredWaveLaneCount` value will need to be added to the PSV3 revision
+ of that struct to accommodate the third preferred value.
+
+#### Device Capability
+
+As a required feature, devices supporting the shader model it ships with are
+ required to respect the wave size restrictions indicated by the wave size range
+ metadata.
+
+## Testing
+
+### Correct Behavior Testing
+
+Verify this compiler output:
+
+* The two parameter overload of `WaveSize` correctly transmits those values to a
+  metadata tuple with a zero third value that is pointed to by the correct tag
+  in the entry point attribute list.
+* The three parameter overload of `WaveSize` transmits those values as well as
+  the third in the same tuple.
+* That the PSV0 `MinimumExpectedWaveLaneCount` and `MaximumExpectedWaveLaneCount`
+  values reflect those provided for the wave size range.
+* That the PSV3 `PreferredWaveLaneCount` value reflects the third preferred
```

Comment:
The runtime doesn't need to know the PreferredWaveLaneCount - nothing actionable for it.  That said, it doesn't hurt to have the information in case somehow it does become useful in the future.  Assuming the cost of the work isn't a big deal.

Reply:
This is a good point.  It's not necessary to expose this value in PSV data.

Reply:
This looks like it is still in the current version

Reply:
Woops. Missed this spot in yesterday's purge. Removed now.